### ISSUE TITLE
New buzzwords Learning To Rank: A Key Information Retrieval Tool for Machine Learning Search

### ISSUE THREAD

**Issue by sudocurse** [sudocurse/reading-list#3039]
Learning To Rank: A Key Information Retrieval Tool for Machine Learning Search<br>
http://ift.tt/2Fjq7vU<br>
<img width="640" height="426" src="http://ift.tt/2qHr2mt" class="webfeedsFeaturedVisual wp-post-image wp-stateless-item" style="display: block; margin: auto; margin-bottom: 20px;" data-image-size="large" data-stateless-media-bucket="cdn.thenewstack.io" data-stateless-media-name="media/2018/01/4de5c135-mar-newhall-226045.jpg">
<p>Ranking isn&rsquo;t just for search engines, or even enterprise search, although it&rsquo;s routinely used by services like Airbnb, Etsy, Expedia, LinkedIn, Salesforce and Trulia to improve search results.</p>
<p>The Learning To Rank (LETOR or LTR) machine learning algorithms <a href="http://ift.tt/2mheGN9" class="ext-link" rel="external nofollow">pioneered first by Yahoo</a>&nbsp;and then <a href="http://ift.tt/2qFhEj2" class="ext-link" rel="external nofollow">Microsoft Research</a>&nbsp;for Bing are proving useful for work such as <a href="http://ift.tt/2mhajle" class="ext-link" rel="external nofollow">machine translation and digital image forensics</a>, computational biology, and <a href="http://ift.tt/2qKol3u" class="ext-link" rel="external nofollow">selective breeding in genetics</a>&nbsp;&mdash; anywhere what you need is a ranked list of items. Ranking also are quickly becoming a cornerstone of digital work.</p>
<p>Beyond document retrieval, LTR is useful for expert search, definition search, collaborative filtering, question answering, keyphrase extraction, document summarization, expert &#64257;nding, anti-web spam, sentiment analysis, and product ratings. LTR methods work for images as well as text (Etsy uses a combination of image and text search to distinguish between products that would get a very similar ranking based on the text description alone, and LTR can <a href="http://ift.tt/2mhanBu" class="ext-link" rel="external nofollow">distinguish higher and lower quality images</a>), or for ranking nodes in a graph. And they&rsquo;re increasingly available to developers, whether you&rsquo;re looking for a Python library or a full-scale enterprise search platform like Apache Solr/Lucene. (<a href="http://ift.tt/2qIgkfB" class="ext-link" rel="external nofollow">Lucene LinkedIn&rsquo;s old search framework</a>; it&rsquo;s no longer used there but was open sourced and is used by sites like Pandora, Soundcloud and Wikipedia. <a href="http://ift.tt/mjFvi2" class="ext-link" rel="external nofollow">Solr</a> is the open source enterprise search platform built on <a href="http://ift.tt/1me1PXO" class="ext-link" rel="external nofollow">Lucene</a>.)</p>
<p>&ldquo;LTR is suitable for pretty much any area where you can find unbiased training data,&rdquo; <a href="http://ift.tt/2BhXIYo" class="ext-link" rel="external nofollow">Ganesh Venkataraman</a>, engineering manager at Airbnb (and former head of relevance at LinkedIn) told the New Stack. &ldquo;It is particularly suitable for cases where there are a large number of features that are harder to hand tune.&rdquo;</p>
<p>&ldquo;LinkedIn uses LTR on almost all of its search verticals &mdash; people search, job search, recruiter search and so on. We have shown this to work very well with sizeable performance gains and have also seen it scale,&rdquo;&nbsp;Venkataraman said. For work such as the job search, the technique ties very closely with another algo, query understanding (QU), which predicts the user intent based on query. For example, if the job search query is &ldquo;google software engineer,&rdquo; LinkedIn will understand that the member is most likely looking for jobs at the COMPANY &ldquo;google&rdquo; with the TITLE &ldquo;software engineer.&rdquo;</p>
<p>Using LTR for matching search queries with relevant ads on Microsoft Bing improved online ad revenue by $4 million in the first two months. Using it to <a href="http://ift.tt/2qJ2NUT" class="ext-link" rel="external nofollow">rank job adverts</a>&nbsp;on LinkedIn didn&rsquo;t just get more people clicking through to look at search results; the search was finding jobs people were looking for, because application rates went up 5 percent in the first three weeks (and up 50 percent by the end of the project). When Bloomberg built an <a href="http://ift.tt/2mhauwU" class="ext-link" rel="external nofollow">LTR plugin for Solr</a>, not only was it a third faster than the hand-built ranking system they&rsquo;d been using before, results were more relevant, and clicks went up by about 10 percent straight away.</p>
<h2>How LTR works</h2>
<p><a href="http://ift.tt/2qIaomx" class="ext-link" rel="external nofollow">Learning To Rank</a>&nbsp;uses <a href="http://ift.tt/2pdwQCl" class="local-link">supervised machine learning</a> to train a model not for the usual single-item classification or prediction, but to discover the best order for a list of items, using features extracted from each item to give it a ranking. It&rsquo;s not looking at the precise score for each item but the relative order &ndash; whether one item is ranks above or below another.</p>
<p>For web results, features might include the domain or PageRank, whether the page title matches the query, the age or popularity of the page, whether it has standard document sections like an introduction or a summary, how long it is and how long people spend reading it, or even whether it will render on a phone screen or just a desktop browser. To make ranking efficient, features need to be extracted automatically; they should be something you can script rather than something you have to code by hand.</p>
<p>The three main LTR techniques are pointwise, pairwise and listwise. Pointwise LTR uses regression or classification to discover the best ranking for individual results. You can think of classification as putting similar documents in the same class and regression as giving similar documents a similar function value, so ranking assigns similar documents similar preferences.</p>
<p>Pairwise LTR uses classification or regression to discover the best order for a pair of items at a time, classifying all the different pairings for items in the list as correctly or incorrectly ranked and working through them to get the ranking for the whole group. There are multiple methods using different techniques for both the cost function and the learning algorithm, including neural networks, random forests, boosting and Support Vector Machines (SVM).</p>
<p>Early pairwise methods like <a href="http://ift.tt/2gCGQCD" class="ext-link" rel="external nofollow">RankNet</a> used a cost function that&rsquo;s about minimizing the number of times a lower rated item is ranked above one with a higher rating and <a href="http://ift.tt/1rIE5iQ" class="ext-link" rel="external nofollow">Stochastic Gradient Descent</a> to train the neural network. <a href="http://ift.tt/2qFhKao" class="ext-link" rel="external nofollow">LambdaRank</a> gets faster, more accurate results using the gradient of that cost (whether an item should move up or down because of its ranking, and by how much), and <a href="http://ift.tt/2mhaDjW" class="ext-link" rel="external nofollow">LambdaMART</a> improves on that using gradient boosted decision trees.</p>
<p>Listwise LTR methods like <a href="http://ift.tt/2qK8pyg" class="ext-link" rel="external nofollow">ListNet rank</a> the whole list rather than working through pairs, using probability models for the cost. For images, using a subset of the list of images <a href="http://ift.tt/2mhaFZ6" class="ext-link" rel="external nofollow">can work better</a>&nbsp;and combining all three cost functions <a href="http://ift.tt/2qJSTT1" class="ext-link" rel="external nofollow">can also improve ranking performance</a>.</p>
<p>The first search engine to use LTR was AltaVista (which was bought by Yahoo) and the key work on LTR was done between 2008 and 2011, with Yahoo and Microsoft releasing training data sets and holding competitions, and the techniques from that era are the ones that are commonly used &mdash; but there&rsquo;s still plenty of on-going research. Venkataraman called &ldquo;the intersection of deep learning and IR (information retrieval)&rdquo; particularly important; recently, he says, &ldquo;Word embeddings (converting query tokens into a higher dimensional space), as well as neural nets, got quite some attention.&rdquo;</p>
<p>There&rsquo;s also research into semi-supervised versions of LTR. Supervised learning relies on well-labeled training data from human experts (like the human judges Google uses to get test scores for changes to its search algorithm), which offers high-quality data &mdash; but it&rsquo;s expensive and only covers a fraction of whatever it is that you want to rank. Analysing search logs and click-through rates is cheaper and works at a much larger scale, but there&rsquo;s often bias in that kind of implicit data (the top results will get more clicks but may not be the best match, and if none of the results are clicked you get no useful information).</p>
<h2><strong>Tools for LTR</strong></h2>
<p>Several LTR tools that were submitted to LTR challenges run by Yahoo, Microsoft and Yandex <a href="http://ift.tt/1PYRw2k" class="ext-link" rel="external nofollow">are available as open source</a>&nbsp;and the <a href="http://dlib.net/ml.html" class="ext-link" rel="external nofollow">Dlib C++ machine learning library</a> includes a tool for training a <a href="http://ift.tt/1gu37x4" class="ext-link" rel="external nofollow">Ranking SVM</a>. The <a href="http://ift.tt/2tXWUAO" class="ext-link" rel="external nofollow">pyltr</a>&nbsp;library is a Python LTR toolkit with ranking models, evaluation metrics and some handy data tools.</p>
<p>For most developers, LTR tools in search tools and services will be more useful. LinkedIn open sourced sample code for building an end-to-end &lsquo;instant&rsquo; search system on <a href="http://ift.tt/2qLZ0pU" class="ext-link" rel="external nofollow">Elasticsearch that uses LTR</a>. The <a href="http://ift.tt/1L5Ojvq" class="ext-link" rel="external nofollow">Azure Search service</a> for building custom search engines has LTR functionality built in for relevance ranking; the custom code Ernst and Young built to create a custom search engine <a href="http://ift.tt/2vqPozx" class="ext-link" rel="external nofollow">for the US tax code</a> is on available <a href="http://ift.tt/2mgtyvo" class="ext-link" rel="external nofollow">as Python scripts and Jupyter notebooks</a>. That&rsquo;s a good example of combining LTR with natural language processing (in this case using Microsoft&rsquo;s LUIS Language Understanding Service API) for extracting features for ranking documents.</p>
<p>Bloomberg&rsquo;s LTR plugin is now built into Solr; it extends the existing RankQuery in the platform and reranks a number of documents using your ranking model (which is deployed as a Solr managed resource) and feature information (so features can be built in Solr with short, simple scripts and rich search options like synonyms and phrase matching still work). That makes sophisticated results ranking available relatively simply, in a popular search platform.</p>
<p class="attribution"><a href="http://bit.ly/2h30eoP" class="ext-link" rel="external nofollow">Microsoft</a> is a sponsor of The New Stack.</p>
<p class="attribution">Feature image&nbsp;by&nbsp;<a href="http://ift.tt/2qJ0p0B" class="ext-link" rel="external nofollow">Mar Newhall</a>&nbsp;on&nbsp;<a href="http://ift.tt/2tazaMI" class="ext-link" rel="external nofollow">Unsplash</a>.</p>
<p>The post <a rel="nofollow" href="http://ift.tt/2Fjq7vU">Learning To Rank: A Key Information Retrieval Tool for Machine Learning Search</a> appeared first on <a rel="nofollow" href="https://thenewstack.io">The New Stack</a>.</p>
<br>
<br>
via Blog &ndash; The New Stack https://thenewstack.io<br>
January 09, 2018 at 06:00AM
